zokrates_core/src/ir/expression.rs

```diff
@@ -122,14 +135,27 @@ impl<T: Field> Sub<LinComb<T>> for LinComb<T> {
 
     fn sub(self, other: LinComb<T>) -> LinComb<T> {
         let mut res = self.0;
+        let mut to_remove = BTreeSet::new();
+
         for (k, v) in other.0 {
-            let new_val = T::zero() - v + res.get(&k).unwrap_or(&T::zero());
-            if new_val == T::zero() {
-                res.remove(&k)
+            let var;
+            {
+                var = res
+                    .iter()
+                    .find(|(id, _)| *id == k)
+                    .map(|(_, v)| v)
+                    .cloned();
+            }
+
+            let new_val = T::zero() - v + var.unwrap_or(T::zero());
+            if new_val != T::zero() {
+                res.push((k, new_val));
             } else {
-                res.insert(k, new_val)
-            };
+                to_remove.insert(k);
+            }
         }
+
+        res.retain(|(id, _)| !to_remove.contains(id));
```

Comment:
So you're filtering out terms that have coefficient 0 at the cost of iterating on the full lincomb. My intuition is that this is close in complexity to reducing to a hashmap, which is what we're trying to avoid. Maybe you have some insights here?

Reply:
Yes, maybe I should build a resulting lincomb while iterating over terms and just not include into it those of them that are resulting to zero.

Reply:
You could filter out `(_42, 0)` which I'm not sure where it would come from, but I don't think you should filter out `[(_42, 1), ...,  (_42, -1)]` as it's the point of this PR to not reduce at each operation and instead only reduce when needed.